Quantity.__array_ufunc__

mhvk opened the pull request:
EDIT (28-Apr-2017): in the end, the numpy implementation is called `__array_ufunc__`.

Following issue #2492, here is an implementation of `__numpy_ufunc__` that at least locally passes all the tests (and allows some previous `xfail`s to be removed!). It also addresses #2483, so hopefully will help set us up for testing against numpy 1.9 in travis.

Notes:
- this requires #2571.
- Rather rough is the implementation of `__numpy_ufunc__` with methods other than `__call__` (e.g., `accumulate`, `reduce`). Right now, I only allow two-argument `ufunc`s that do not change the unit (this is in the new `scales_and_result_unit` function, which is in `quantity_helper`). Do comment if this is wrong! Or if it is true for only some of those methods.
- I let overrides of some methods depend on the numpy version; let me know if this is done the right way.

mhvk commented:
I removed the milestone as there are still bugs on the numpy side that need to be resolved (and hopefully will before numpy 1.10).

mhvk commented:
Another rebase of the `__numpy_ufunc__`, where I now use it also for wrapping other numpy methods and functions, thus reducing code duplication. The coverage failure is a red herring: it is mostly because code now only is included if numpy versions are sufficiently ancient, and this code is, of course, not tested with the more recent numpy version that coverage uses.

mdboom commented:
How does this relate to #2948?

mhvk commented:
#2948 is a follow-up -- it requires `__numpy_ufunc__` to be present (as does #3554, obviously).

mhvk commented:
Rebased yet again.  Still waiting for https://github.com/numpy/numpy/pull/4815 as well as the resolution of `astropy.io` errors with numpy-dev (see #3584).

eteq commented:
@mhvk - came across this while following a trail of some older PRs... is it still relevant?  it looks like numpy/numpy#4815 was closed at some point...

mhvk commented:
Yes, it is still relevant, although the whole `__numpy_ufunc__` implementation has been pushed off to 1.12 (maybe even under a different name).  But let's keep it around for now...

eteq commented:
OK, thanks for clearing that up.  Will assign it to upstream-required until further notice then.

mhvk commented:
Numpy (finally) gained support for overriding ufuncs -- https://github.com/numpy/numpy/pull/8247 -- which means that unit conversions will no longer lead to having to recalculate the result. So, time to revive this PR -- in time still for 2.0!

mhvk commented:
Small to dos:
- [ ] Changelog entry
- [ ] remove the "dev" in `NUMPY_LT_1_13` (needed for to ensure this is tested correctly right now)

mhvk commented:
Travis failures are all unrelated (though the sphinx one may be real, not just URL timeouts).

bsipocz commented:
@mhvk - I've restarted travis. One of the failing jobs  (No8 in the matrix) may be related to changes here, as the timeout occurred in the quantity array tests.

mhvk commented:
@bsipocz - good call, it seems to have failed again. I'll investigate.

mhvk commented:
Hmm, strange, it does pass locally. I restarted the test again.

bsipocz commented:
@mhvk - How long should we support numpy 1.7? Maybe it's time to remove it? (Though it might be better to understand why the timeout occurs, but getting rid of np 1.7 for v2.0 sounds good enough workaround for me)

mhvk commented:
Hmm, there was actually a real underlying problem, in which nominally better checks on casting were done incorrectly, which caused segmentation faults on numpy 1.7. It should now been fixed...

(Of course, we may still want to stop supporting numpy 1.7. It would remove quite a bit of cruft.)

mhvk commented:
Rebased on #5998, since otherwise the numpy-dev tests fail for not supporting `np.positive`.

mhvk commented:
I think this is ready for review; @adrn, @astrofrog?

To help this a bit: much of the changes are just to refactor out checks that were done in `__array_prepare__` to helper functions in `quantity_helper`; these are then used in `__array_ufunc__` as well.

To get a sense of the benefit of not calculating the wrong result first:
```
q = np.linspace(0, 360, 10000) * u.deg
%timeit np.sin(q)
# current master: 1000 loops, best of 3: 619 µs per loop
# with this PR and numpy-dev: 1000 loops, best of 3: 290 µs per loop
```

p.s. The documentation failure is unrelated (see #5997); the coverage error is a false positive (the substantial `__array_ufunc__` addition is counted as extra code, but not tested by the numpy <1.13 run that determines coverage).

mhvk commented:
Another comparison:
```
q = np.linspace(0, 360, 10000) * u.deg
a = np.deg2rad(q.value)
%timeit np.sin(q)
# 1000 loops, best of 3: 289 µs per loop
%timeit np.sin(a)
# 1000 loops, best of 3: 247 µs per loop
```

adrn commented:
Loving what I see, but need to look closer to do a proper review. Will try to get to it during #pyastro17 next week.

bsipocz commented:
ping @adrn for a review now at #pyastro17

adrn commented on astropy/units/quantity.py:
```diff
@@ -372,123 +356,39 @@ def __array_prepare__(self, obj, context=None):
+            # Quantity with an inconsistent unit (e.g., not angular for Angle),
+            # and that we can handle the type (e.g., that we are not int when
+            # float is required).
+            check_output(obj, result_unit, (args + tuple(
```
Do you mind explaining this with a little more comments? I'm not following the need for `float16` and `int8`.

mhvk replied:
`check_output` will internally get a common dtype for all those args. Now suppose I have two integer quantities `q1` and `q2` (these are difficult to construct, but still allowed...), with units `km` and `m` and try `q1 += q2` -- in this case, the converter will just be a division by 1000, so the if test fails, and the `float16` ensures that the `check_output` will test that the output argument can handle `float` (it is `float16` as the minimum length float, just in case `q1` and `q2` are also `float16`).

I must admit the case for `int8` is less clear -- I think I just added it as a small integer alternative.

To be honest, looking at this again, I'm also not sure it isn't better to just go with with what was above (except replace `float` with `float16` since there is no reason to disallow that). Though, then again, I think I even added a test for the above but with `q1` in `m` and `q2` in `km`, in which case integers are OK...

mhvk commented:
The current errors in -dev and -rc2 are due to an upstream numpy bug: https://github.com/numpy/numpy/pull/9112

bsipocz commented:
@mhvk - Could you rebase? With the upstream numpy PR being merge, this passed the restarted builds already.

mhvk commented:
Rebased, and fixed two regressions introduced by the modeling unit work -- a bug each in modeling and in my implementation. I think this is ready to go in....

adrn commented on astropy/units/quantity_helper.py:
```diff
@@ -361,10 +369,252 @@ def helper_twoarg_floor_divide(f, unit1, unit2):
+        if unit is None:
+            raise TypeError("Cannot store non-quantity output{0} in {1} "
+                            "instance".format(
+                                ("from {0} function".format(function.__name__)
```
space before "from"?

adrn commented on astropy/units/quantity_helper.py:
```diff
@@ -361,10 +369,252 @@ def helper_twoarg_floor_divide(f, unit1, unit2):
+            raise UnitTypeError(
+                "Cannot store output with unit '{0}'{1} "
+                "in {2} instance.  Use {3} instance instead."
+                .format(unit, ("from {0} function".format(function.__name__)
```
again, space before "from"?

mhvk commented:
OK, fixed those. Since that's all untested strings, skipping CI and merging...

pllim commented:
After 3 years! 🎆

mhvk commented:
Indeed, it still needed a lot of work and to-and-fro on the numpy side....

taldcroft commented:
Congratulations @mhvk and thank you so much for your persistence and numpy cross-project work!  🎉
